src/transactions/OfferExchange.cpp

```diff
@@ -110,98 +117,72 @@ exchangeV3(int64_t wheatReceived, Price price, int64_t maxWheatReceive,
     return result;
 }
 
-LoadBestOfferContext::LoadBestOfferContext(Database& db, Asset const& selling,
-                                           Asset const& buying)
-    : mSelling(selling)
-    , mBuying(buying)
-    , mDb(db)
-    , mBatchIterator(mBatch.end())
-{
-    loadBatchIfNecessary();
-}
-
-void
-LoadBestOfferContext::loadBatchIfNecessary()
-{
-    if (mBatchIterator == mBatch.end())
-    {
-        mBatch.clear();
-        OfferFrame::loadBestOffers(5, 0, mSelling, mBuying, mBatch, mDb);
-        mBatchIterator = mBatch.begin();
-    }
-}
-
-OfferFrame::pointer
-LoadBestOfferContext::loadBestOffer()
-{
-    return (mBatchIterator != mBatch.end()) ? *mBatchIterator : nullptr;
-}
-
-void
-LoadBestOfferContext::eraseAndUpdate()
-{
-    ++mBatchIterator;
-    loadBatchIfNecessary();
-}
-
-OfferExchange::OfferExchange(LedgerDelta& delta, LedgerManager& ledgerManager)
-    : mDelta(delta), mLedgerManager(ledgerManager)
-{
-}
-
 OfferExchange::CrossOfferResult
-OfferExchange::crossOffer(OfferFrame& sellingWheatOffer,
-                          int64_t maxWheatReceived, int64_t& numWheatReceived,
-                          int64_t maxSheepSend, int64_t& numSheepSend)
+OfferExchange::crossOffer(
+    LedgerState& ls, std::shared_ptr<LedgerEntryReference> sellingWheatOffer,
+    int64_t maxWheatReceived, int64_t& numWheatReceived,
+    int64_t maxSheepSend, int64_t& numSheepSend)
 {
     assert(maxWheatReceived > 0);
     assert(maxSheepSend > 0);
 
-    // we're about to make changes to the offer
-    mDelta.recordEntry(sellingWheatOffer);
+    // Note: These must be copies, not references.
+    Asset sheep = sellingWheatOffer->entry()->data.offer().buying;
```

Comment:
Looks very complicated to get into the "buing" and "selling" fields, any chance to make that simpler?